### ISSUE TITLE
Timing violations

### ISSUE THREAD

**Issue by WizardAtLarge** [battosai30/OneWire_Workaround_0.0.9#1]
1) Please refer to the following apps for specs on the signal timing for bit-write and bit-read of 1-wire devices. The values you propose do not meet the minimum requirements
https://www.maximintegrated.com/en/app-notes/index.mvp/id/126
https://www.maximintegrated.com/en/app-notes/index.mvp/id/162

2) To properly write the 1-wire driver, if the micro is sufficiently fast (as the m0 is), both the "standard" and "overdrive" speeds for 1-wire devices can be implemented. App note 126 discusses this in more detail.

3) Your implementation of  "#define TIMING_C 1.2-10" could be refined to something like
#define TIMING_C  1 - (DELAY_MICROSECONDS_FUNCTION_OVERHEAD)
where DELAY_MICROSECONDS_FUNCTION_OVERHEAD = (Clock cycles overhead of the delayMicroseconds function) / (clock frequency of micro)

For example, if the delayMicroseconds function has 80 clock cycle overhead and the controller is at 48MHz, then DELAY_MICROSECONDS_FUNCTION_OVERHEAD = 80/48 or approx 2 usecs and thus
TIMING_C <== 1-2

But ... if the desired delay of "delayMicroseconds" is less than DELAY_MICROSECONDS_FUNCTION_OVERHEAD, then the number computed is less than 0 which is an error ==> this is the case in line 177 of your function "OneWire::reset"
-->   delayMicroseconds(2 * TIMING_C)
        ==> 2 * TIMING_C = 2*1.2-10 = 2-10 = -8 == 0xFFFFFFFFFFFFFFF8 (unsigned long)

4) A solution that would keep delayMicroseconds() at the system level could be:
   a) Create a macro that compensates for both scenarios (delays greater than) and (less than or equal) and use this in place of the calls to "delayMicroseconds"
-OR-
   b) Write an intermediate function named something like delayMicrosecondsOneWire() that takes into account all delays (including the call to delayMicrosecondsOneWire) and passes the compensated value to delayMicroseconds(); That is, 
    i) add private function delayMicrosecondsOneWire() to onewire.cpp
    ii) replace all calls in onewire.cpp from delayMicroseconds() to delayMicrosecondsOneWire()
   iii) Make delayMicrosecondsOneWire() dependent on parameters defined in the variant.cpp and variant.h files

**Comment by battosai30** [comment]
Hi,

TIMING_C etc are just testing values I forgot to take off. 
Basically I did not published this correction to be a "real" correction. I just works, and the original one not. So it's just a workaround but it deserves of course a real work. 

To get to it I spend a lot of time playing with logic analyzer and compare with working plateform (ESP32 + DS18B20). So now I should at least clean up the code and I know that read_bit seemed to be the main source of problem. Switching from output to input mode seems too long and the DS118b20 latch the line too soon.

**Comment by WizardAtLarge** [comment]
ESP32 Arduino 1-wire library had issues - do not know if Stickbreaker's changes have been merged into the main branch or not. ESP32 Arduino runs under FreeRTOS and the timing jitters, when a higher priority task interrupts, can be problematic. Stickbreaker (aka Chuck Todd) patch works. 

For timing, the "timing min/max window" specs need to be met. This is CRITICAL during the 1-wire search operation (https://www.maximintegrated.com/en/app-notes/index.mvp/id/187), (https://www.maximintegrated.com/en/app-notes/index.mvp/id/4104) and (https://www.maximintegrated.com/en/app-notes/index.mvp/id/148). These are good apps to read and understand.

You need to look at the analog waveforms - if you were using a digital logic analyzer, you may not be observing the long rise times when the controller releases the driver (1-wire is open-collector based). For 100KHz, the 1-wire timing spec is conservative such that 100m equivalent cable lengths will work reliably.

When testing DS18B20's on 1 meter or less cable, there is little delay due to capacitance and signal propagation.   In one installation, I have (10) DS18B20 sensors equally spaced along a 30m cable. I used 18AWG wire because 24AWG wire had too much capacitance and resistance for reliable/consistent search results. 

Open-source code has issues and must be used "at risk". For the bug that you found (driving signal low THEN changing port direction) is one such bug. It may be a bug for STM32L0 devices - but may not for other microcontrollers. What is happening (per your scope plots):
1) a '0' is written to the port => signal goes LOW
2) Port reconfiguration (from OUTPUT to OUTPUT) note: compiler does not optimize this out
  * port is tri-stated ==>external pull-up resistor of the 1-wire circuit pulls signal HIGH
  * port is configured for OUTPUT ==> signal goes LOW

In summary, thanx for finding the bug in the 1-wire code for STM32Lxxx devices! By improving the code via determining the code overhead/latency, adjusting the timing accordingly and meeting the timing specs for 1-wire communications, you got a sweet patch that GrumpyOldPizza may want to merge into his trunk.
